### ISSUE TITLE
bug with combineLatest() ?

### ISSUE THREAD

**Issue by Lakedaemon** [ReactiveX/RxJava#4377]
These last days, I was burned by a somewhat unexpected behavior of combineLatest (RxJava 1.1.9) :
When I used with 2 observables on different threads (because of subscribeOn), it emits nothing 

Am I doing something wrong ? or is this a bug ? as designed ?

``` java
import javafx.util.Pair;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import rx.Observable;
import rx.schedulers.Schedulers;

@RunWith(JUnit4.class)
public class Bug {
    @Test
    void test() throws Exception {
        StringBuilder builder = new StringBuilder();
        // this one works as expected
        builder.append("first test\n");

        Observable.combineLatest(
            Observable.just(1, 2, 3, 4, 5, 6, 7), 
            Observable.just(2, 5, 7, 9, 2, 4, 5), 
            (a, b) -> new Pair(a, b)).subscribe(it -> builder.append(it.toString()));

        // this one doesn't  emit anything
        builder.append("second test\n");
         Observable.combineLatest(
            Observable.just(1, 2, 3, 4, 5, 6, 7).subscribeOn(Schedulers.io()), 
            Observable.just(2, 5, 7, 9, 2, 4, 5), 
            (a, b) -> new Pair(a, b)).subscribe(it -> builder.append(it.toString()));

        // this one doesn't  emit anything
        builder.append("third test\n");
        Observable.combineLatest(
            Observable.just(1, 2, 3, 4, 5, 6, 7).observeOn(Schedulers.io()).map(it -> it), 
            Observable.just(2, 5, 7, 9, 2, 4, 5), 
            (a, b) -> new Pair(a, b)).subscribe(it -> builder.append(it.toString()));

        throw new Exception(builder.toString());
    }
}
```

**Comment by akarnokd** [comment]
Your test quits before those other asynchronous `combineLatest`s can even run.  Even if they'd run, the `StringBuilder` would get corrupted due to concurrent access.

**Comment by Lakedaemon** [comment]
Yes, you are right. I wrote crappy tests that can't fulfill their purpose.

I have an issue in one android app with combineLatest that never emits anything though the 2 observables that feeds it (one observable created through Observable.create() <-- this might be the issue (bad implementation, backpressure not implemented but needed ?), and one observable created through a publishSubject with startWith(1 value)) each emit something. 

I'll go back to investigating what goes wrong in my app...

**Comment by Lakedaemon** [comment]
It's weird. 
I fixed (or I think that I did) the Observable.onCreate() : 
a) I wasn't using onCompleted() at the end, 
b) I also added an onBackpressureBuffer() after it
c) Also, it goes through a scan operator that emits  a data collector that feeds combineLatest()
I also made sure that I'm returning a different collector (in case combineLatest tests if the result is the same).... 
yet, though the scan observable emits lots of items on a 20 seconds span... the combineLatest() only emits 1 item at the very end, when the scan operator is finished... really weird.
I'll try to investigate further...

**Comment by Lakedaemon** [comment]
I got it working. Actually, it was point c that was causing my issue.
In the scan operator I was always returning the same accumulator and I guess that there is a test in the combineLatest operator to avoid emitting a value that hasn't changed.

Returning a copy of the accumulator with a slightly different id fixed the equal result and combineLatest suddenly started emitting mor items.  

Maybee this could/should be added in the documentation of combineLatest if it isn't already there.

**Comment by Lakedaemon** [comment]
I'm reopening this issue because my problem might still come from a combineLatest bug

Finding it in my app is quite tricky and I wrongly thought that it might come from an equal test.

Apparently, if I use Observable.combineLatests(observable1, observable2) it fails
but if I switch the observables Observable.combineLatests(observable2, observable1), it succeeds.

So, I'm thinking that It might be a timing issue or a bug in combineLatest.
I haven't much time to code but I'll invstiguate further and try to write a minimal test showing the bug

**Comment by Lakedaemon** [comment]
I'm still observing nasty issues with combineLatest. 

I'm wondering if the problem lies with the way it requests items from my 2 observables

When my issue appears : one observable emits like crazy (like 290 items on a io thread) and the other observable (an Observable.just(int) on the current thread) waits and emit when the other observable is finished. Then, the combineLatest emits it's first item  when it should have emitted like 290 items instead

sadly, the combineLatest code is really complicated and I would be at a loss to check if this really happens But I'm pretty sure that the issue lies with combineLatest

**Comment by akarnokd** [comment]
CombineLatest prefetches 128 elements from each source. If your source is synchronous, that means it will run through it and you'll get its very last item only. You rarely want to combine with synchronous sources due to this property. Don't you need `zip` instead?

**Comment by Lakedaemon** [comment]
I have my issue with an Observable.just(cursor) (that I'll replace with a PublishSubject<Int>.startWith(cursor) once things work out and an asynchronous Observable<Collector> that emits 290 times (on the novel where the issue appears) after each page it collects.

with logs in 3 places (2 inside map after the 2 observables and 1 in the combining function), I observed that the asynchronous observable was emitting 290 items and the other observable was somehow waiting till it was finished to emit it's single item and then the combination function was called once.

zip wouldn't work because I'm spliting a novel into 291 pages, but as it is quite long (20s), I also emit the position where the reader is reading and if pagination has already done that page, it can return it immediately.

the publishSubject is used in case the reader swipes to the next/previous page or if he jumps to a completely different page. That way, I can split pages in the background but serve him page immediately (if he jumps at the end, I just paginate 1 page from his chosen position while paginating continues from start...)

**Comment by Lakedaemon** [comment]
I finally found what was wrong. There is no issue with combineLatest (as expected).

I was using subscribeOn() in a foolish attempt to make some observable run either on the UI thread or on a Schedulers.io() thread... and this had those surprising result (related to locks ? races ? contention ?). 
I guess that it prevented the subscription of Observable.just() till the other observable was finished which surprised me

My observable created through Observable.create() was fine (but now that I understand things better, I should probably use SyncOnSubscribe inside). 

Now, to make a custom observable source work on Schedulers.io(), I'm using 

Observable.just(Unit).observeOn(Schedulers.io()).flatMap(myObservable)

It is certainly overkill (and a waste) to use a flatMap to make my observable work on another thread, but it's the first and only solution I found at the moment.

My conclusions : subscribeOn(), when misused, can introduce some very nasty issues, very hard to debug. Now, I know.

Many thanks for your kind answers and your awesome contributions to RxJava, you rock !
